yarn-project/end-to-end/src/e2e_2_rpc_servers.test.ts

```diff
@@ -0,0 +1,148 @@
+    await aztecRpcServerB?.stop();
+  });
+
+  const expectBalance = async (wallet: Wallet, owner: AztecAddress, expectedBalance: bigint) => {
+    const contractWithWallet = new ZkTokenContract(contractWithWalletA.address, wallet);
+    const [balance] = await contractWithWallet.methods.getBalance(owner).view({ from: owner });
+    logger(`Account ${owner} balance: ${balance}`);
+    expect(balance).toBe(expectedBalance);
+  };
+
+  const expectsNumOfEncryptedLogsInTheLastBlockToBe = async (numEncryptedLogs: number) => {
+    const l2BlockNum = await aztecNode.getBlockHeight();
+    const encryptedLogs = await aztecNode.getLogs(l2BlockNum, 1, LogType.ENCRYPTED);
+    const unrolledLogs = L2BlockL2Logs.unrollLogs(encryptedLogs);
+    expect(unrolledLogs.length).toBe(numEncryptedLogs);
+  };
+
+  const expectUnencryptedLogsFromLastBlockToBe = async (logMessages: string[]) => {
+    const l2BlockNum = await aztecNode.getBlockHeight();
+    const unencryptedLogs = await aztecNode.getLogs(l2BlockNum, 1, LogType.UNENCRYPTED);
+    const unrolledLogs = L2BlockL2Logs.unrollLogs(unencryptedLogs);
+    const asciiLogs = unrolledLogs.map(log => log.toString('ascii'));
+
+    expect(asciiLogs).toStrictEqual(logMessages);
+  };
+
+  const deployContract = async (initialBalance: bigint, from: AztecAddress) => {
+    logger(`Deploying L2 contract...`);
+    const tx = ZkTokenContract.deploy(aztecRpcServerA, initialBalance, from).send();
+    const receipt = await tx.getReceipt();
+    contractWithWalletA = new ZkTokenContract(receipt.contractAddress!, walletA);
+    contractWithWalletB = new ZkTokenContract(receipt.contractAddress!, walletB);
+    await tx.isMined(0, 0.1);
+    const minedReceipt = await tx.getReceipt();
+    expect(minedReceipt.status).toEqual(TxStatus.MINED);
+    logger('L2 contract deployed');
+    return contractWithWalletA;
+  };
+
+  it('transfers fund from user A to B via RPC Server A followed by transfer from B to A via RPC Server B', async () => {
+    const initialBalance = 987n;
+    const transferAmount1 = 654n;
+    const transferAmount2 = 323n;
+
+    // Deploy ZkToken contract with walletA
+    await deployContract(initialBalance, userA);
+
+    // Add account B pub key and partial address to wallet A
+    const [accountBPubKey, accountBPartialAddress] = (await aztecRpcServerB.getPublicKeyAndPartialAddress(userB))!;
+    await aztecRpcServerA.addPublicKeyAndPartialAddress(userB, accountBPubKey, accountBPartialAddress);
+    // Add account A pub key and partial address to wallet B
+    const [accountAPubKey, accountAPartialAddress] = (await aztecRpcServerA.getPublicKeyAndPartialAddress(userA))!;
+    await aztecRpcServerB.addPublicKeyAndPartialAddress(userA, accountAPubKey, accountAPartialAddress);
+
+    // Add zkToken to rpc server B
+    await aztecRpcServerB.addContracts([
+      {
+        abi: ZkTokenContract.abi,
+        address: contractWithWalletA.address,
+        portalContract: EthAddress.ZERO,
+      },
+    ]);
```

Comment:
I'm surprised this is needed. Aren't we publishing all contract deployment info to the chain, so the synchroniser takes care of this? What are we missing?

Reply:
Good point. It throws an unknown contract error when it's not there. I will investigate whether it's an issue with Synchroniser.

Reply:
As discussed in the sprint planning, this is a correct behavior.